File: src/main/java/com/mattiaswikstrom/hyper/Space.java
package com.mattiaswikstrom.hyper;

import java.io.*;
import java.util.*;

public final class Space implements Serializable
{
    public Trig trig;
    Collection<Element> elements;

    public Iterator getElementIterator()
    {
        return elements.iterator();
    }

    public void draw(javafx.scene.canvas.GraphicsContext g)
    {
        for (Iterator i = elements.iterator(); i.hasNext(); )
        {
            Element e = (Element) i.next();
            e.draw(g);
        }
    }

    public Space(double k)
    {
        trig = new Trig(k);
        elements = new LinkedList<Element>();
    }

    public double getCurvature()
    {
        return trig.getCurvature();
    }

    void changeCurvature(double k)
    {
        trig = new Trig(k);
    }

    public Length distance(Point p1, Point p2)
    {
        return trig.distance(p1.d, p2.d, Angle.diff(p2.a, p1.a));
    }

    public void rotate(Angle angle)
    {
        for (Iterator i = elements.iterator(); i.hasNext(); )
        {
            Element e = (Element) i.next();
            e.rotate(angle);
        }
    }

    public void moveforward(Length dist)
    {
        for (Iterator i = elements.iterator(); i.hasNext(); )
        {
            Element e = (Element) i.next();
            e.moveForward(dist);
        }
    }
}

File: hyper-web/src/Space.ts
import { Length  } from "./Length.js";
import { Angle  } from "./Angle.js";
import { Trig  } from "./Trig.js";
import { Point  } from "./Point.js";
import { SpaceElement  } from "./SpaceElement.js";

export class Space {
    trig: Trig;
    elements: SpaceElement[];

    constructor(k: number, _elements:SpaceElement[]) {
        this.trig = new Trig(k);
        this.elements = _elements;
    }

    getElementIterator(): Iterator<SpaceElement> {
        return this.elements[Symbol.iterator]();
    }

    draw(g: CanvasRenderingContext2D): void {
        for (let e of this.elements) {
            e.draw(g);
        }
    }

    getCurvature(): number {
        return this.trig.getCurvature();
    }

    changeCurvature(k: number): void {
        this.trig = new Trig(k);
    }

    distance(p1: Point, p2: Point): Length {
        return this.trig.distance(p1.getDistance(), p2.getDistance(), Angle.diff(p2.getAngle(), p1.getAngle()));
    }

    rotate(angle: Angle): void {
        for (let e of this.elements) {
            e.rotate(angle);
        }
    }

    moveforward(dist: Length, space: Space): void {
        for (let e of this.elements) {
            e.moveForward(dist, space);
        }
    }
}
